Read inactive vlans

Izhmash opened the pull request:
- Addresses #921
- Effects PowerConnect, Nexus, and N3000 switches.
- `get_port_networks()` now returns every VLAN in the ranges rather than the ones at the end of the ranges.
- I am releasing this now for review to get feedback before writing the deployment tests to test more than two trunking VLANs

zenhack commented on hil/ext/switches/n3000.py:
```diff
@@ -44,14 +41,13 @@ class DellN3000(Switch):
-                              lambda v: 0 < v <= 4093,
-                              Use(str)),
+            'dummy_vlan': schema.And(schema.Use(int),
+                                     lambda v: 0 < v and v <= 4093,
```
Why get rid of the `0 < v <= 4093` shorthand?

Also, while I don't necessary have an objection to qualifying all of the variables from the `schema` module, I don't understand how it's related to this patch?

Izhmash replied:
I'll clean this PR up, I'm not sure how this change snuck in there.

Izhmash replied:
Actually this should be it for style-related things, the rest should be functional. I'm not sure why the diff is appearing a bit strangely in `_dell_base`

zenhack commented on hil/ext/switches/_dell_base.py:
```diff
@@ -35,37 +36,37 @@ def disable_native(self, vlan_id):
+            non_native_list = []
+            # Get native vlan then remove junk if native not None
+            native_vlan = v['Trunking Native Mode VLAN'].strip()
+            if (native_vlan != 'none'):
```
parens aren't necessary.

zenhack commented on hil/ext/switches/_dell_base.py:
```diff
@@ -35,37 +36,37 @@ def disable_native(self, vlan_id):
+            native_vlan = v['Trunking Native Mode VLAN'].strip()
+            if (native_vlan != 'none'):
+                native_vlan = ''.join(c for c in native_vlan
+                                      if c not in badchars)
```
This is somewhat subjective, but I personally find python's if expression really hard to read. Maybe just use filter and a lambda?

naved001 replied:
personally I find filter and lambdas hard to read. I find what @Izhmash has done to be more readable.

zenhack commented on hil/ext/switches/_dell_base.py:
```diff
@@ -35,37 +36,37 @@ def disable_native(self, vlan_id):
+                native_vlan = None
+            # Get other vlans and parse out junk if not None
+            trunk_vlans = v['Trunking VLANs Enabled'].strip()
+            if (trunk_vlans != 'none'):
```
Same thing; no need for parens in an if.

Izhmash commented:
@zenhack I've added some documentation for the Dell switch, if the docs there are good I can add them to the other drivers too.

Overall, my `get_port_networks()` iterates through ports' port configurations and gets every native + non-native trunking VLAN mentioned in the configuration.  The majority of the code here is string parsing, which some logic that branches off depending on if the switch reports VLANs or if it reports 'none'.

naved001 commented:
> I don't understand how this fixes the bug? in #921 the discussion settled on the problem probably being the newline, not the inactive string. I don't understand where the solution is here?

I think @Izhmash ended up rewriting the whole method so that in addition to reading inactive vlans, it also parses the vlan range correctly.

naved001 commented on hil/ext/switches/_dell_base.py:
```diff
@@ -36,36 +37,46 @@ def disable_native(self, vlan_id):
+                # create comma-separated list to use parse_vlans()
+                non_natives = ','.join(non_natives)
+                non_native_list = parse_vlans(non_natives)
+            for v in (non_native_list):
```
haha writing in Go did this to @Izhmash

zenhack replied:
I don't understand how? what construct in Go was confusing him?

Izhmash replied:
No worries, just a typo

zenhack commented:
Ah, okay, I understand what's going on now.

Izhmash commented on hil/ext/switches/_dell_base.py:
```diff
@@ -36,36 +37,46 @@ def disable_native(self, vlan_id):
+            native_vlan = v['Trunking Native Mode VLAN'].strip()
+            if native_vlan != 'none':
+                native_vlan = ''.join(c for c in native_vlan
+                                      if c not in badchars)
```
I'll give that a go, that sounds a lot better.

Izhmash commented:
> You've got a fair amount of copypasta from one driver to another. Ideally we'd factor out the common bits.

The n3000 and the PowerConnect code here is the same, should I move the function I wrote to `_dell_base`?

Izhmash commented:
> You've got a fair amount of copypasta from one driver to another. Ideally we'd factor out the common bits.
The n3000 function I wrote is the same as the PowerConnect's; I can have them use the same code.  I'll look into factoring out the Nexus code too.

Izhmash commented:
So I've removed the n3000 port networks codes and created a unified on in `_dell_base`.  The n3000 and PowerConnect are very similar; only small changes in key word ordering and none representations got in theway.

I've tried to make it more readable by separating some of the code out.

Izhmash commented:
My latest commit allows the deployment tests to run successfully.  Tested for Nexus and PowerConnect.

Izhmash commented:
N3000 deployment tests working as well.

Izhmash commented:
I've added a deployment test that tests the discovery of multiple (>2) networks on one port using `get_port_networks`.

naved001 commented on hil/ext/switches/_dell_base.py:
```diff
@@ -36,37 +37,74 @@ def disable_native(self, vlan_id):
+                        network_list.append(('vlan/native', int(native_vlan)))
+                else:
+                    # appropriate to append native vlan to list
+                    network_list.append(('vlan/native', int(native_vlan)))
```
we could probably collapse the 2 if statements into 1:
```python
if hasattr(self.switch, 'dummy_vlan') and int(native_vlan) == int(self.switch.dummy_vlan):
    native_vlan = None
else:
    network_list.append(('vlan/native', int(native_vlan)))
````

that way we don't have 2 repeated else blocks either.

naved001 commented on hil/ext/switches/_dell_base.py:
```diff
@@ -36,37 +37,74 @@ def disable_native(self, vlan_id):
+            result[k] = network_list
         return result
 
+    def _make_vlan_list(self, dirty_list):
```
this method could be outside the class (it doesn't need `self` at all)

Izhmash commented:
@naved001 I fixed those up.

naved001 commented on hil/ext/switches/_dell_base.py:
```diff
@@ -36,35 +37,63 @@ def disable_native(self, vlan_id):
+        'Trunking VLANs Enabled': ' 23 - 25'
+
+        Return:
+        {<hil.model.Port object at 0xDEADBEEF0123>: [('vlan/native', '3')],
```
deadbeef

zenhack commented on tests/deployment/multi_networks.py:
```diff
@@ -0,0 +1,152 @@
+
+        The above is super vague; unfortunately the setup operations are very
+        slow, so it makes a huge difference to do everything in one pass. See
+        the comments in-line to understand exactly what is being tested.
```
We've now got this exact docstring copied in 3 places. I'm worried about how much copypasta I've been seing in patches to the test suite lately...

zenhack commented on tests/deployment/multi_networks.py:
```diff
@@ -0,0 +1,152 @@
+            network_create_simple('net-1', 'anvil-nextgen')
+            network_create_simple('net-2', 'anvil-nextgen')
+            network_create_simple('net-3', 'anvil-nextgen')
+            network_create_simple('net-4', 'anvil-nextgen')
```
This seems like it should be  a loop (same below).

Izhmash commented:
@zenhack I think I addressed your comments.  Also I'd like to test this again since it's been a little while, I'll put in a comment when that's done

naved001 commented on tests/deployment/multi_networks.py:
```diff
@@ -63,12 +63,11 @@ def create_multi_nets():
 
             nodes = self.collect_nodes()
 
+            deferred.apply_networking()
```
this shouldn't be here.

Izhmash replied:
oops I'll remove that

Izhmash commented:
Deployment tests are still passing

zenhack commented on tests/deployment/multi_networks.py:
```diff
@@ -0,0 +1,147 @@
+
+    def test_multi_networks(self):
+        """A suite of tests to ensure that multiple networks can relate
+        to a single port. Previously, there was an issue with
```
Would be good to add a reference to the specific Github issue.

Izhmash replied:
Should be all set now
